presto-orc/src/main/java/com/facebook/presto/orc/CachingOrcDataSource.java

```diff
@@ -0,0 +1,165 @@
+/*
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.facebook.presto.orc;
+
+import com.facebook.presto.orc.metadata.StripeInformation;
+import com.google.common.annotations.VisibleForTesting;
+import com.google.common.collect.ImmutableMap;
+import com.google.common.primitives.Ints;
+import io.airlift.slice.FixedLengthSliceInput;
+import io.airlift.slice.Slices;
+import io.airlift.units.DataSize;
+
+import java.io.IOException;
+import java.util.List;
+import java.util.ListIterator;
+import java.util.Map;
+
+import static com.google.common.base.Verify.verify;
+import static java.util.Objects.requireNonNull;
+
+public class CachingOrcDataSource
+        implements OrcDataSource
+{
+    private static final long MAX_CACHE_LENGTH = new DataSize(8, DataSize.Unit.MEGABYTE).toBytes();
+
+    private final OrcDataSource dataSource;
+    ListIterator<StripeInformation> stripeIterator;
+
+    private long cachePosition;
+    private int cacheLength;
+    private byte[] cache;
+
+    public CachingOrcDataSource(OrcDataSource dataSource, List<StripeInformation> stripes)
+    {
+        this.dataSource = requireNonNull(dataSource, "dataSource is null");
+        this.stripeIterator = requireNonNull(stripes, "stripes is null").listIterator();
+    }
+
+    public static OrcDataSource wrapIfTinyStripes(OrcDataSource dataSource, List<StripeInformation> stripes)
+    {
+        if (stripes.size() < 3) {
```

Comment:
Why did we choose 3 here?  I would assume that we would want to do this for even a single stripe since it will coalesce all reads into a single read.